[misc] Add json processor for exporting questionnaires to importable XMLs

acrowthe opened the pull request:
JSON to XML export process:

1. Grab the form's URL from the admin dashboard:
`http://localhost:8080/content.html/admin/Questionnaires/CompactSectionsTest`


2. Remove `/content.html/admin`, add all the needed export tags to the end (`.deep.-identify.importable.json`):
`http://localhost:8080/Questionnaires/CompactSectionsTest.deep.-identify.importable.json`
4.  Save the json (right click -> "Save As" or "Save Page As" depending on what browser you are using)
     - For simplicity, save the file to `cards/Utilities/JSON-to-XML/` with a name of your choice (`CompactSectionsTest.json` in this case)

OR **Replace steps 2&3 with clicking the `Export as JSON` entry in the questionnaire dropdown menu** 


5. In terminal, navigate to `cards/Utilities/JSON-to-XML/` and run the converter (editing the file names as appropriate):
`python3 json_to_xml.py CompactSectionsTest.json > CompactSectionsTest.xml`
     - The resulting XML will be in `cards/Utilities/JSON-to-XML/`

To quickly test the generated XML, you can deploy it to a running instance with:

```
curl 'http://localhost:8080/Questionnaires/' -u admin:admin -F ':operation=import' -F ':contentType=xml' -F ":content=@FileName.xml" -F ':name=QuestionnaireId' -F ':replace=true'
```
(Update port, FileName and QuestionnaireId as needed)

This process will not lead to identical XMLs from the original source XMLs:
- Any default properties will be included in the newly exported file (eg. `paginate` will be on all questionnaire nodes as it defaults to `False` when originally imported)
- Ordering may not be the same

acrowthe committed:
[misc] Add json processor for exporting questionnaires to importable xmls

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/JsonToXmlProcessor.java:
```diff
@@ -0,0 +1,84 @@
+                case "jcr:lastModified":
+                case "jcr:lastModifiedBy":
+                case "jcr:uuid":
+                    // Skip most jcr nodes. Can't skip all as jcr:primaryType is still needed
```
I think we should explicitly allow `jcr:primaryType` and remove anything else starting with `jcr:`. Other properties that may show up later are `jcr:isCheckedOut`, `jcr:lastCheckedOut`, `jcr:baseVersion`, `jcr:predecessors`, `jcr:versionHistory`, and potentially `jcr:activity` and `jcr:mergeFailed`.

marta- replied:
Agreed

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/JsonToXmlProcessor.java:
```diff
@@ -0,0 +1,84 @@
+public class JsonToXmlProcessor implements ResourceJsonProcessor
+{
+    @Reference
+    private FormUtils formUtils;
```
This doesn't seem to be used.

marta- replied:
why is this needed?

marta- commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/JsonToXmlProcessor.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    public String getName()
+    {
+        return "jsontoxml";
```
Works great but I don't love the name of the processor. Something like `-edithistory` comes to mind (remove all data about who contributed to the resource and when), but better names are welcome.

sdumitriu replied:
I think it's OK to have a processor that outputs something for the import without having to combine different small parts, and while at the moment this happens to just remove the properties that deal with versioning and history, other properties may be removed at later times. On the contrary, I would like to include some of the things that other processors do to not have such a long list of selectors, so this could also remove any (json) properties starting with `@` and `sling/`. Or something better that I've been meaning to do for a while but is not urgent, add inter-processor dependencies so that this one automatically pulls in `-identify` and `nolinks` and whatever else it needs.

How about `importable` as the name?

marta- replied:
I'm ok with `importable` and I like the idea of this processor pulling the other processors

marta- commented:
Should we restrict this processor to questionnaire resources only? i think applying this to other resources like forms might not be useful and have side effects.

marta- committed:
[misc] Enable JSON export for questionnaires in the admin UI

sashaandjic reviewed (approved):
tested

acrowthe committed:
[misc] Add json processor for exporting questionnaires to importable XMLs
- Rename processor from jsonToXmlProcessor to ImportableProcessor
- Add more functionality to the processor so it isn't dependant on as many
  other processors
- Expand jcr: property removal to handle other jcr: properties that may be present

acrowthe commented:
Renamed to `.importable` and removed the requirement for the `.simple` and `.nolinks` processors
